File: Assignment-2/Question-4/q4.cpp
#include <iostream>
#include <fstream>
#include <ctime>
using namespace std;

int main(){
    ofstream outfs(results.txt)
    int n[6] = {16,64,256,1024,4096,16384};
    int iter = 0;
    outfs << "m = 13421772800" << endl;
    while(iter < 6){
        int a,b,x;
        int** M = new int* [n[iter]];
        for(int i = 0; i < n[iter]; i++){
            M[i] = new int[n[iter]];
        }

        for(int i = 0; i < n[iter]; i++){
            for(int j = 0; j < n[iter]; j++){
                M[i][j] = 2;
            }
        }

        size_t m = 13421772800;
        clock_t t;
        t = clock();

        for(int i = 0; i < m; i++){
            x = rand() % 100 + 1;
            a = rand() % n[iter];
            b = rand() % n[iter];
            M[a][b] = M[a][b] + x
        }
        t = clock() - t;
        outfs << "n = " << n[iter] << " takes"
              << float(t) / CLOCKS_PER_SEC << " seconds" << endl;
        iter++;
    }
    return 0;
}

File: Assignment-2/Question-4/q4.py
import random
import time

n = [16, 64, 256, 1024, 4096, 16384]
iter = 0

with open('results.txt', 'w') as outfs:
    outfs.write("m = 13421772800\n")
    
    while iter < 6:
        a, b, x = 0, 0, 0
        m = [[2 for _ in range(n[iter])] for _ in range(n[iter])]

        m_size = 13421772800
        t_start = time.time()

        for i in range(m_size):
            x = random.randint(1, 100)
            a = random.randint(0, n[iter] - 1)
            b = random.randint(0, n[iter] - 1)
            m[a][b] = m[a][b] + x

        t_end = time.time() - t_start
        outfs.write(f"n = {n[iter]} takes {t_end} seconds\n")
        iter += 1
